### PR TITLE
Improved <br> and whitespace fix

### PR DESCRIPTION
**Pull request by pkruithof**
### TL;DR

This fixes #12 in a different, less destructive way.

### Longer version 

By putting the html through the `DOMDocument` class, a lot of changes are made to it, which in turn can lead to different Markdown, especially when the html is invalid in all sorts of ways (which is rather common on the web, ask browser vendors :wink: ).

In my case I had html that didn't properly close tags:

```html
<b>header<b><br>
Some text
<br><br>

<b>Another header
<br><br>
Some more text
```

Which before resulted in the following Markdown:

```md
**header**
Some text

**Another header**

Some more text

```

Now it does something like this:

```md
**header**
Some text

****Another header****

Some more text

```

This keeps on growing on each header. Now, of course the html is wrong, but I still think the resulting Markdown should represent what you'd see if you loaded that html [in a web browser][1]:

```md
**header
Some text**

**Another header**

Some more text

```

[1]: http://codepen.io/anon/pen/pvBvgy

This is also according to the CommonMark spec, which allows for a single newline in inline elements ([test][2]), but not for 2 ([test][3])

[2]: http://spec.commonmark.org/dingus/?text=*foo%20%20%0Abar*%0A
[3]: http://spec.commonmark.org/dingus/?text=*foo%0A%0Abar*%0A

I've updated the `parse()` method to fix both problems in #12, while adhering to the rules explained above. This is also updated in the test suite (which by the way could really use a rewrite that doesn't rely on trailing spaces, since many editors/IDE's trim those).

### CODE REVIEW

**Comment by tzi** [comment]
Thanks for this very well explained PR.

Your new code fit more with the legacy code!
I don't know if it's good... but it will help us to limit side effects ;)

Could you, please:

 * Move the code you added in the `parse()` function to fix spaces and breaks, in one separate function. It will help other to understand & maintain this code.
 * Add a test corresponding to your PR example

The trailing spaces in the test cases are also boring me.
What do you propose?

Cheers,
Thomas.

**Comment by pkruithof** [comment]
> I don't know if it's good... but it will help us to limit side effects ;)

What is does basically, is when the parser is in inline context it looks for leading/trailing spaces in the text inside that element and when found, it moves that whitespace outside of the tag. I've also documented this in the function, along with examples.

> Move the code you added in the parse() function to fix spaces and breaks, in one separate function. It will help other to understand & maintain this code.

Done :+1: 

> Add a test corresponding to your PR example

The test I added is similar to the example I described in this PR.

> The trailing spaces in the test cases are also boring me.
> What do you propose?

What I would do is, make all markdown strings in the test case single-lined. That way you can still have trailing spaces, and they would be preserved because of the ending quote. In case of a multi-line Markdown example you could split it up in multiple single-line strings, like this:

```php
$data['next-to-bold-and-br']['md'] = "**McHale**  \n" .
"\n" .
"\n" .
"  1. Bird" . 
"  2. Magic"
;
```

**Comment by tzi** [comment]
Nice improvments!

Thanks for this PR ;)

Cheers,
Thomas.

**Comment by tzi** [comment]
BTW, I'm open to a PR that split up multiple lines strings in the test cases.

Cheers,
Thomas.

**Comment by pkruithof** [comment]
You're welcome, could you maybe release a new version with this change?

**Comment by tzi** [comment]
The release is ready 
https://github.com/Elephant418/Markdownify/releases/tag/v2.1.9

See you!
Thomas.

**Comment by tzi** [comment]
The multi lines string in test cases are split up, since 9f453df.

Cheers,
Thomas.

**Comment by pkruithof** [comment]
Nice! :+1:
